### ISSUE TITLE
Unsafe pointer from string

### ISSUE THREAD

**Issue by iskiselev** [sq/JSIL#860]
Looks like JSIL doesn't support resolving pointer from string. Here is test case:
```
using System;

static class Program {
    public static void Main () {
        string str = "qww";
        unsafe
        {
            fixed (char* sPtr = str)
            {
                *sPtr = 'a';
            }
        }

        Console.WriteLine(str);
    }
}
```

Translation result:
```
    var sPtr = JSIL.UntranslatableInstruction("Implicit conversion to pointer: Untranslatable Expression Conversion of expression '@<var System.String str>' to pointer");
    sPtr.set(("a").charCodeAt(0));
    $T00().WriteLine("qww");
```

**Comment by iskiselev** [comment]
It would be very good, if at least read operations will work.

**Comment by iskiselev** [comment]
Looks like it will be pretty easy to implement for read operations. Will do it.
In any case, mutate string through a pointer is not best idea even in .Net.

**Comment by kg** [comment]
Yeah, PinAndGetPointer just needs to copy the string's chars into a temporary utf-16 buffer.

**Comment by iskiselev** [comment]
Will it be OK if we create Pointer with null as MemoryRange and source string as view (I've already done it)? Or there will be problems if we'd like to cast pointer afterward?

**Comment by kg** [comment]
No, that's no good.

**Comment by iskiselev** [comment]
What should we do on write through such pointer attempts? Right now I throw NotImplementedException. Or should we allow write to such temporary array (waithout writing it back to source string)?

**Comment by kg** [comment]
I would allow the write and not represent it in the source string. It's not possible to mutate a string instance in JS anyway.

While mutating a pinned string in C# might work it's a terrible idea so I'm okay with not implementing it. If it's needed, we could make it work by implementing the unpin portion of fixed blocks and using that to flush the change to the string somehow... but the string's object reference would change, so it's still not gonna work.

**Comment by iskiselev** [comment]
Additional notes: get pointer to char[] element doesn't work also.
One more thing to test - cast pointer to byte* or int* (as sizeof(char)=2).
```
using System;

static class Program {
    public unsafe static void Main(string[] args)
    {
        string str = "qww";
        unsafe
        {
            fixed (char* sPtr = str.ToCharArray())
            {
                RunMe(sPtr, str.Length);
            }
        }

        Console.WriteLine(str);
    }

    private unsafe static void RunMe(char* sPtr, int length)
    {
        char* endPtr = sPtr + length;
        byte* bPtr = (byte*)sPtr;
        while (sPtr < endPtr)
        {
            Console.WriteLine(*sPtr);
            Console.WriteLine(*bPtr);
            Console.WriteLine(*(bPtr + 1));

            sPtr++;
            bPtr += 2;
        }
    }
}
```

**Comment by kg** [comment]
char is tricky since it's a string instance right now. at some point we
should switch them over to ints but the lack of boxing prevents it
currently. same goes for bool.

**Comment by iskiselev** [comment]
Could you explain why lack of boxing prevents us from moving chars to Numbers? Right now I don't see why it is so.

**Comment by kg** [comment]
char and boolean would lose all type information when cast if unboxed, which poses some difficulties. there's also some other stuff that needs to happen when indexing into strings, etc. i think once boxing/wrapping works the rest can be sorted out.

**Comment by markusjohnsson** [comment]
That is what I do in [cil.js](https://github.com/markusjohnsson/cil.js) and it works great. The type is implicitly known after the unboxing and always provided when boxing occurs.
